### ISSUE TITLE
"comments" for polygon nesting

### ISSUE THREAD

**Issue by mdsumner** [edzer/sp#64]
@rsbivand is the comment meaning documented somewhere?  My take is:  

- the attribute matches length of each `@Polygons` 
- 0 (island)
- n >=1 (is a hole in Polygon n)

so the maximum value n could take (for feature 1)  is `length(x@polygons[[1]]@Polygons) - 1`?

I used to find this extremely opaque, but I understand the `comment()` mechanism now and its relation to attributes (and how this is applied in the C code). 

(any chance of migrating rgdal out to github? I know it's an unwelcome overhead atm, but I think the extra accessibility would be helpful)

**Comment by rsbivand** [comment]
No, not there. The hidden `comment` attribute has existed in base R for a long time. I used it to "piggy-back" an SF-conformant vector of integers in a space-separated string in 2010/2011 to provide a way of using GEOS in **rgeos** - while S3 class objects can add attributes or components without breaking the API, S4 objects cannot modify slots. So Polygons objects needed something extra to indicate which rings were exterior rings, and for interior rings, which exterior ring they belonged to. In **sp**, this has worked adequately for almost 10 years, so I thought that the same trick might be used to "piggy-back" a WTK2 string onto an S4 `CRS` object using a comment. Comments also accompany the commented object through `.Call()`, so can be retrieved in C/C++. 

Once this is done and restores transformation accuracy lost in the shift from GDAL2/PROJ5 to GDAL3/PROJ6, a blog/article needs writing, possibly with an appendix showing how the use of `comment()` and `comment<-()` works in this case. Users should never need to see the workings, WKT2 is close to illegible for humans (like XML).

Your concern about custom projections should be addressable by instantiating `CRS` objects with WKT2 comment populated by modifying `checkCRSArgs()` which is run by `CRS()`, using importFromProj4() in GDAL, which is needed by GDAL for multiple drivers.

I will not move **rgdal** to github, and as long as I maintain the package, it will stay that way. All new uses should migrate to **sf**/**stars**, so encouraging PRs and new functionality in a package that has served its purpose is inappropriate, I feel. The interfaces to PROJ and GDAL are hard to maintain in **rgdal** and **rgeos**, and **sf**'s use of **Rcpp** provides for much easier support and maintenance. 

Hope this clarifies.

**Comment by mdsumner** [comment]
It doesn't, the polygon comments are not documented so I'll just assume my take is correct until I find otherwise.

**Comment by rsbivand** [comment]
Sorry, misunderstood your point. The use of comments to provide a template for SF conformity is documented fairly fully in ASDAR, 2.ed., pp. 47-48. I'll create a new issue to add this to **sp**.
